File: mcmt-tracking/mcmt-tracking/src/mcmt_track_python/mcmt_track_python/mcmt_single_tracker_node.py
# basic Python frameworks
import cv2
import math
import time
import numpy as np

# ROS2 libraries
import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter
from rclpy.clock import Clock

# import message interface types
from mcmt_msg.msg import SingleDetectionInfo
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError

# local imported code
from mcmt_track_python.mcmt_track_utils import TrackPlot, scalar_to_rgb


class SingleTrackerNode(Node):
	"""
	Single camera tracking node, for processing the single camera tracks' features
	"""
	def __init__(self):
		super().__init__("SingleTrackerNode")
		
		# declare video parameters
		self.bridge = CvBridge()
		self.index = None
		self.timer = time.time()
		
		# declare and get mcmt ROS2 parameters
		self.declare_mcmt_parameters()
		self.get_mcmt_parameters()
		
		# get camera parameters
		cap = cv2.VideoCapture(self.index)
		self.frame_w = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
		self.frame_h = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
		self.scale_factor = math.sqrt(self.frame_w ** 2 + self.frame_h ** 2) / math.sqrt(848 ** 2 + 480 ** 2)
		self.aspect_ratio = self.frame_w / self.frame_h
		self.fps = int(cap.get(cv2.CAP_PROP_FPS))

		if self.frame_w * self.frame_h > (self.max_frame_w * self.max_frame_h):
			self.frame_w = 1920
			self.frame_h = int(1920 / self.aspect_ratio)
			self.scale_factor = math.sqrt(self.frame_w ** 2 + self.frame_h ** 2) / math.sqrt(848 ** 2 + 480 ** 2)
		
		self.recording = cv2.VideoWriter(self.output_vid_path,
										cv2.VideoWriter_fourcc(*'mp4v'),
										self.fps, (self.frame_w, self.frame_h))
		cap.release()

		# data from detector
		self.good_tracks = []
		self.frame_count = 0
		self.total_num_tracks = None
		self.frame = None
		self.next_id = None

		self.origin = np.array([0, 0])
		self.track_plots_ids = []
		self.track_plots = []

		# initialize plotting parameters
		self.plot_history = 200
		self.colours = [''] * self.plot_history
		for i in range(self.plot_history):
			self.colours[i] = scalar_to_rgb(i, self.plot_history)
		self.font = cv2.FONT_HERSHEY_SIMPLEX
		self.font_scale = 0.5

		# create subscriber to the topic name "mcmt/detection_info"
		topic_name = "mcmt/detection_info"
		self.detect_sub = self.create_subscription(
			SingleDetectionInfo, topic_name, self.track_callback, 10)
		# prevent unused variable warning
		self.detect_sub

		# create publisher to the topic name "mcmt/track_image"
		topic_name = "mcmt/track_image"
		self.track_pub = self.create_publisher(Image, topic_name, 1000)
		# prevent unused variable warning
		self.track_pub

	
	def declare_mcmt_parameters(self):
		"""
		This function declares our mcmt software parameters as ROS2 parameters
		"""
		self.declare_parameter('IS_REALTIME')
		self.declare_parameter('VIDEO_INPUT')
		self.declare_parameter('FRAME_WIDTH')
		self.declare_parameter('FRAME_HEIGHT')
		self.declare_parameter('OUTPUT_VIDEO_PATH')
		self.declare_parameter('OUTPUT_CSV_PATH')


	def get_mcmt_parameters(self):
		"""
		This function gets the mcmt parameters from the ROS2 paramters
		"""
		# get camera parameters
		self.is_realtime = self.get_parameter('IS_REALTIME').value
		
		if self.is_realtime:
			self.index = int(self.get_parameter('VIDEO_INPUT').value)
		else:
			self.index = self.get_parameter('VIDEO_INPUT').value
		
		self.max_frame_w = self.get_parameter('FRAME_WIDTH').value
		self.max_frame_h = self.get_parameter('FRAME_HEIGHT').value
		self.output_vid_path = self.get_parameter('OUTPUT_VIDEO_PATH').value
		self.output_csv_path = self.get_parameter('OUTPUT_CSV_PATH').value

	
	def track_callback(self, msg):
		"""
		callback function to process the camera's tracks' features
		"""
		start_timer = time.time()

		# get DetectionInfo message
		try:
			self.frame = self.bridge.imgmsg_to_cv2(msg.image, desired_encoding="passthrough")
		except CvBridgeError as e:
			print(e)
		
		# get goodtracks list
		self.total_num_tracks = len(msg.goodtracks_id)
		self.good_tracks = []
		for i in range(self.total_num_tracks):
			self.good_tracks.append([msg.goodtracks_id[i], msg.goodtracks_x[i], msg.goodtracks_y[i]])
		print(f"Time take to get message: {time.time() - self.timer}")

		# get track feature variable for each track
		for track in self.good_tracks:
			track_id = track[0]
			centroid_x = track[1]
			centroid_y = track[2]

			if track_id not in self.track_plots_ids:  # First occurrence of the track
				self.track_plots_ids.append(track_id)
				self.track_plots.append(TrackPlot(track_id))

			track_plot = self.track_plots[self.track_plots_ids.index(track_id)]
			track_plot.update((centroid_x, centroid_y), self.frame_count)
			track_plot.calculate_track_feature_variable(self.frame_count, self.fps)

		# plot trackplots
		for track_plot in self.track_plots:
			idxs = np.where(np.logical_and(track_plot.frameNos > self.frame_count - self.plot_history,
											track_plot.frameNos <= self.frame_count))[0]
			for idx in idxs:
				cv2.circle(self.frame, (track_plot.xs[idx] - self.origin[0], track_plot.ys[idx] - self.origin[1]),
							3, self.colours[track_plot.frameNos[idx] - self.frame_count + self.plot_history - 1][::-1], -1)
			if len(idxs) != 0:
				cv2.putText(self.frame, f"ID: {track_plot.id}",
							(track_plot.xs[idx] - self.origin[0], track_plot.ys[idx] - self.origin[1] + 15),
							self.font, self.font_scale, (0, 0, 255), 1, cv2.LINE_AA)
				if track_plot.track_feature_variable.size != 0:
					cv2.putText(self.frame, f"Xj: {np.mean(track_plot.track_feature_variable):.3f}",
								(track_plot.xs[idx] - self.origin[0], track_plot.ys[idx] - self.origin[1] + 30),
								self.font, self.font_scale, (0, 255, 0), 1, cv2.LINE_AA)
		
		# get timer
		end_timer = time.time()
		print(f"Trackplot process took: {end_timer - start_timer}")

		# show and save video tracking frame
		self.frame_count += 1
		self.imshow_resized_dual("Detection", self.frame)
		self.recording.write(self.frame)
		self.timer = time.time()
		cv2.waitKey(1)

	
	def imshow_resized_dual(self, window_name, img):
		"""
		Function to resize and enlarge tracking frame
		"""
		aspect_ratio = img.shape[1] / img.shape[0]

		window_size = (int(1280), int(1280 / aspect_ratio))
		img = cv2.resize(img, window_size, interpolation=cv2.INTER_CUBIC)
		cv2.imshow(window_name, img)

File: mcmt-tracking/mcmt-tracking/src/mcmt_track/src/mcmt_single_tracker_node.cpp
/**
 * @file mcmt_single_tracker_node.cpp
 * @author Dr Sutthiphong Srigrarom (Spot), spot.srigrarom@nus.edu.sg
 * @author Mr Niven Sie, sieniven@gmail.com
 * @author Mr Seah Shao Xuan, seahshaoxuan@gmail.com
 * @author Mr Lau Yan Han, sps08.lauyanhan@gmail.com
 * 
 * This code is conceptualised, created and published by the SPOT-IT 3D team
 * from the Department of Mechanical Engineering, Faculty of Engineering 
 * at the National University of Singapore. SPOT-IT 3D refers to the 
 * Simultaneous Positioning, Observing, Tracking, Identifying Targets in 3D.
 * This software utilizes a multi-camera surveillance system for real-time 
 * multiple target tracking capabilities. This software capability is highly
 * applicable for monitoring specific areas, and some use cases include monitoring 
 * airspaces, traffic junctions, etc.
 * 
 * This file is part of the SPOT-IT 3D repository and can be downloaded at:
 * https://github.com/sieniven/spot-it-3d
 * 
 * This file contains the declarations of the functions primarily used in the
 * tracking pipeline. These functions interact with the key classes CameraTracks 
 * and TrackPlots which are essential to the tracking process. These classes may 
 * be found at mcmt_track_utils.cpp. This code also contains the McmtSingleTrackerNode 
 * class which handles the subscription from our ROS2 DDS-RTPS ecosystem.
 */

// local header files
#include <mcmt_track/mcmt_single_tracker_node.hpp>

// standard package imports
#include <stdlib.h>
#include <stdio.h>
#include <iostream>
#include <fstream>
#include <csignal>
#include <chrono>
#include <math.h>
#include <memory>
#include <map>
#include <set>
#include <algorithm>
#include <functional>
#include <eigen3/Eigen/Dense>

using namespace mcmt;

McmtSingleTrackerNode::McmtSingleTrackerNode() : Node("SingleTrackerNode") {

	node_handle_ = std::shared_ptr<rclcpp::Node>(this, [](::rclcpp::Node *) {});
	declare_parameters();
	get_parameters();
	RCLCPP_INFO(this->get_logger(), "Initializing Mcmt Single Tracker Node");

	// get camera parameters
	if (is_realtime_ == true) {
		cap_ = cv::VideoCapture(std::stoi(video_input_));
	} else {
		cap_ = cv::VideoCapture(video_input_);
	}

	frame_w_ = int(cap_.get(cv::CAP_PROP_FRAME_WIDTH));
	frame_h_ = int(cap_.get(cv::CAP_PROP_FRAME_HEIGHT));
	scale_factor_ = (sqrt(pow(frame_w_, 2) + pow(frame_h_, 2))) / (sqrt(pow(848, 2) + pow(480, 2)));
	aspect_ratio_ = frame_w_ / frame_h_;
	fps_ = int(cap_.get(cv::CAP_PROP_FPS));

	// if video frame size is too big, downsize
	downsample_ = false;
	if ((frame_w_ * frame_h_) > (FRAME_WIDTH_ * FRAME_HEIGHT_)) {
		downsample_ = true;
		frame_w_ = FRAME_WIDTH_;
		frame_h_ = int(FRAME_WIDTH_ / aspect_ratio_);
		scale_factor_ = (sqrt(pow(frame_w_, 2) + pow(frame_h_, 2))) / (sqrt(pow(848, 2) + pow(480, 2)));
	}

	// intialize video writer;
	recording_ = cv::VideoWriter(output_vid_path_, cv::VideoWriter::fourcc('M','P','4','V'), fps_, 
		cv::Size(frame_w_, frame_h_));
	cap_.release();

	// initialize frame count
	frame_count_ = 0;

	// initialze plotting parameters
	plot_history_ = 200;
	font_scale_ = 0.5;

	cumulative_tracks_ = std::shared_ptr<CameraTracks>(new CameraTracks(0));

	process_detection_callback();

}

/**
 * This function creates subscription to the detection info publisher from the McmtSingleDetectorNode
 * and contains the main pipeline for the tracker node callback. The pipeline includes:
 * 1. Processing of new tracks
 * 2. Plotting of info on each camera's frames
 */
void McmtSingleTrackerNode::process_detection_callback()
{
	// set qos and topic name
	auto qos = rclcpp::QoS(rclcpp::KeepLast(10));
	topic_name_ = "mcmt/detection_info";
	detection_sub_ = this->create_subscription<mcmt_msg::msg::SingleDetectionInfo> (
		topic_name_,
		qos,
		[this](const mcmt_msg::msg::SingleDetectionInfo::SharedPtr msg) -> void
		{
			// get start time
			auto start = std::chrono::system_clock::now();	

			// declare tracking variables
			std::shared_ptr<cv::Mat> frame_;
			std::vector<std::shared_ptr<GoodTrack>> good_tracks_;

			// process detection info
			process_msg_info(msg, frame_, good_tracks_);

			int track_id, centroid_x, centroid_y, size;
			for (auto & track : good_tracks_) {
				track_id = track->id;
				centroid_x = track->x;
				centroid_y = track->y;
				size = track->size;

				std::vector<int> location;
				location.push_back(centroid_x);
				location.push_back(centroid_y);

				if (cumulative_tracks_->track_plots_.find(track_id) == cumulative_tracks_->track_plots_.end()) {
					cumulative_tracks_->track_plots_[track_id] = std::shared_ptr<TrackPlot>(new TrackPlot(track_id));
					debug_messages.push_back("New target ID " +  std::to_string(track_id));
				}

				cumulative_tracks_->track_plots_[track_id]->update(location, size, frame_count_);
			}

			print_frame_summary(good_tracks_);
			annotate_frames(frame_, cumulative_tracks_);
			graphical_UI(frame_, cumulative_tracks_);

			// get trackplot process time
			auto end = std::chrono::system_clock::now();
			std::chrono::duration<double> elapsed_seconds = end - start;
			std::cout << "Trackplot process took: " << elapsed_seconds.count() << "s\n";		

			// show cv window
			std::string window_name = "Annotated";
			imshow_resized_dual(window_name, *frame_.get());
			recording_.write(*frame_.get());

			frame_count_ += 1;

			cv::waitKey(1);
		}
	);
}

/**
 * This function processes the detection messsage information
 */
void McmtSingleTrackerNode::process_msg_info(mcmt_msg::msg::SingleDetectionInfo::SharedPtr msg,
	std::shared_ptr<cv::Mat> & frame,
	std::vector<std::shared_ptr<GoodTrack>> & good_tracks) {
	
	// get camera frame from message
	frame = std::shared_ptr<cv::Mat>(
		new cv::Mat(
			msg->image.height, msg->image.width, encoding2mat_type(msg->image.encoding),
			const_cast<unsigned char *>(msg->image.data.data()), msg->image.step));

	// get good tracks
	int total_num_tracks = msg->goodtracks_id.size();

	for (int i = 0; i < total_num_tracks; i++) {
		auto good_track = std::shared_ptr<GoodTrack>(new GoodTrack());
		good_track->id = msg->goodtracks_id[i];
		good_track->x = msg->goodtracks_x[i];
		good_track->y = msg->goodtracks_y[i];
		good_track->size = msg->goodtracks_size[i];
		good_tracks.push_back(good_track);
	}
}


void McmtSingleTrackerNode::print_frame_summary(std::vector<std::shared_ptr<GoodTrack>> good_tracks_) {

	std::cout << "SUMMARY OF FRAME " << frame_count_ << std::endl;
	std::cout << "Tracks: ";
	for (auto good_track : good_tracks_) {
		std::cout << "(" << good_track->id << ") | ";
	}
	std::cout << std::endl;
}

void McmtSingleTrackerNode::annotate_frames(std::shared_ptr<cv::Mat> frame_, std::shared_ptr<CameraTracks> cumulative_tracks_) {

	
	// draw tracks on opencv GUI to monitor the detected tracks
	// loop through each camera frame
	std::map<int, std::shared_ptr<mcmt::TrackPlot>>::iterator track;

	cv::putText(*frame_.get(), "Frame Count " + std::to_string(frame_count_), cv::Point(20, 25),
		cv::FONT_HERSHEY_SIMPLEX, font_scale_ * 0.85, cv::Scalar(255, 0, 0), 2, cv::LINE_AA);
	
	// loop through every track plot
	if (cumulative_tracks_->track_plots_.empty() == false) {
		for (track = cumulative_tracks_->track_plots_.begin(); track != cumulative_tracks_->track_plots_.end(); track++) {
			if ((frame_count_ - track->second->lastSeen_) <= fps_) {
				
				cv::Point2i rect_top_left((track->second->xs_.back() - (track->second->size_.back())), 
														(track->second->ys_.back() - (track->second->size_.back())));
	
				cv::Point2i rect_bottom_right((track->second->xs_.back() + (track->second->size_.back())), 
																(track->second->ys_.back() + (track->second->size_.back())));
	
				
				cv::rectangle(*frame_.get(), rect_top_left, rect_bottom_right, colors[track->second->id_ % 10], 2);

				cv::Scalar status_color;
				if (track->second->lastSeen_ == frame_count_) {
					status_color = cv::Scalar(0, 255, 0);
				} else {
					status_color = cv::Scalar(0, 0, 255);
				}
				
				// get last frames up till plot history (200)
				shown_indexes_.clear();

				for(int j = track->second->frameNos_.size() - 1; j >= 0; j--) {
					if (track->second->frameNos_[j] > (frame_count_ - plot_history_)) {
						shown_indexes_.push_back(j);
					} else {
						break;
					}
				}

				for (auto & idx : shown_indexes_) {
					int color_idx = track->second->frameNos_[idx] - frame_count_ + plot_history_ - 1;
					double alpha = 0.5 + (double) color_idx / 400;
					double beta = 1 - alpha;
					cv::Vec3b pixelColor = (*frame_.get()).at<cv::Vec3b>(track->second->ys_[idx], track->second->xs_[idx]);

					cv::circle(*frame_.get(), cv::Point(track->second->xs_[idx], track->second->ys_[idx]), 3,
						cv::Scalar((int) (pixelColor[0] * beta + (colors[track->second->id_ % 10][0] * alpha)),
									(int) (pixelColor[1] * beta + (colors[track->second->id_ % 10][1] * alpha)),
									(int) (pixelColor[2] * beta + (colors[track->second->id_ % 10][2] * alpha))), -1);
				}

				// put ID on opencv GUI
				if (shown_indexes_.empty() == false) {
					cv::putText(*frame_.get(), "ID: " + std::to_string(track->second->id_).substr(0,4), 
						cv::Point(rect_top_left.x + 20, rect_top_left.y - 5), cv::FONT_HERSHEY_SIMPLEX,
						font_scale_, colors[track->second->id_ % 10], 1, cv::LINE_AA);
				}
				
				cv::circle(*frame_.get(), cv::Point(rect_top_left.x + 5, rect_top_left.y - 10), 5, status_color, -1);	

			}

			// if (track->second->check_stationary()) {
			// 	cv::putText(*frame_.get(), "S", cv::Point(track->second->xs_.back(), track->second->ys_.back() - 40),
			// 			cv::FONT_HERSHEY_SIMPLEX, font_scale_ * 2, cv::Scalar(255, 0, 0), 2, cv::LINE_AA);
			// }
		}
	}
}

void McmtSingleTrackerNode::graphical_UI(std::shared_ptr<cv::Mat> frame_, std::shared_ptr<CameraTracks> cumulative_tracks_) {
	
	// Summary box
	cv::rectangle(*frame_.get(), cv::Point(190, 860), cv::Point(1000, 900), cv::Scalar(220,220,220), -1);
	cv::rectangle(*frame_.get(), cv::Point(190, 860), cv::Point(1000, 900), cv::Scalar(110,110,110), 4);
	int spacing = 0;
	int drones_on_screen = 0;
	if (cumulative_tracks_->track_plots_.empty() == false) {
		for (auto track = cumulative_tracks_->track_plots_.begin(); 
			track != cumulative_tracks_->track_plots_.end(); track++) {
			if ((frame_count_ - track->second->lastSeen_) <= fps_) {
				drones_on_screen++;
				cv::putText(*frame_.get(), "ID: " + std::to_string(track->second->id_).substr(0,4), cv::Point(210 + spacing, 890), cv::FONT_HERSHEY_SIMPLEX,
						font_scale_ * 1.5, colors[track->second->id_ % 10], 2, cv::LINE_AA);
				spacing += 150;	
			}
		}
	}

	// Notification box
	cv::rectangle(*frame_.get(), cv::Point(20, 920), cv::Point(1000, 1060), cv::Scalar(200,200,200), -1);
	int num_of_messages = 4;
	spacing = 0;
	for (int i = 0; i < num_of_messages && i < debug_messages.size(); i++, spacing -= 30) {
		cv::putText(*frame_.get(), debug_messages[debug_messages.size() - 1 - i], cv::Point(40, 1040 + spacing), 
						cv::FONT_HERSHEY_SIMPLEX, font_scale_ * 1.5, cv::Scalar(0,0,0), 2, cv::LINE_AA);
	}

	// Targets box
	cv::rectangle(*frame_.get(), cv::Point(20, 780), cv::Point(170, 900), cv::Scalar(220,220,220), -1);
	cv::rectangle(*frame_.get(), cv::Point(20, 780), cv::Point(170, 900), cv::Scalar(110,110,110), 4);
	cv::putText(*frame_.get(), "TARGETS", cv::Point(45, 805), cv::FONT_HERSHEY_SIMPLEX,
				font_scale_ * 1.5, cv::Scalar(0,0,0), 2, cv::LINE_AA);
	cv::putText(*frame_.get(), std::to_string(drones_on_screen), cv::Point(60, 885), cv::FONT_HERSHEY_SIMPLEX,
				font_scale_ * 6, cv::Scalar(0,0,0), 6, cv::LINE_AA);
}

void McmtSingleTrackerNode::imshow_resized_dual(std::string & window_name, cv::Mat & img) {
	cv::Size img_size = img.size();

	double aspect_ratio = (double) img_size.width / (double) img_size.height;

	cv::Size window_size;
	window_size.width = 1080;
	window_size.height = (int) (1080.0 / aspect_ratio);
	
	cv::Mat img_;
	cv::resize(img, img_, window_size, 0, 0, cv::INTER_CUBIC);
	cv::imshow(window_name, img_);
}

/**
 * This function declares our mcmt software parameters as ROS2 parameters.
 */
void McmtSingleTrackerNode::declare_parameters() {
	// declare ROS2 video parameters
	this->declare_parameter("IS_REALTIME");
	this->declare_parameter("VIDEO_INPUT");
	this->declare_parameter("FRAME_WIDTH");
	this->declare_parameter("FRAME_HEIGHT");
	this->declare_parameter("OUTPUT_VIDEO_PATH");
	this->declare_parameter("OUTPUT_CSV_PATH");
}

/**
 * This function gets the mcmt parameters from the ROS2 parameters
 */
void McmtSingleTrackerNode::get_parameters() {
	// get video parameters
	IS_REALTIME_param = this->get_parameter("IS_REALTIME");
	VIDEO_INPUT_param = this->get_parameter("VIDEO_INPUT");
	FRAME_WIDTH_param = this->get_parameter("FRAME_WIDTH");
	FRAME_HEIGHT_param = this->get_parameter("FRAME_HEIGHT");
	OUTPUT_VIDEO_PATH_param = this->get_parameter("OUTPUT_VIDEO_PATH");
	OUTPUT_CSV_PATH_param = this->get_parameter("OUTPUT_CSV_PATH");

	// initialize video parameters
	is_realtime_ = IS_REALTIME_param.as_bool();
	video_input_ = VIDEO_INPUT_param.as_string();
	FRAME_WIDTH_ = FRAME_WIDTH_param.as_int();
	FRAME_HEIGHT_ = FRAME_HEIGHT_param.as_int();
	output_vid_path_ = OUTPUT_VIDEO_PATH_param.as_string();
	output_csv_path_ = OUTPUT_CSV_PATH_param.as_string();
}

int McmtSingleTrackerNode::encoding2mat_type(const std::string & encoding) {
	if (encoding == "mono8") {
			return CV_8UC1;
	} else if (encoding == "bgr8") {
			return CV_8UC3;
	} else if (encoding == "mono16") {
			return CV_16SC1;
	} else if (encoding == "rgba8") {
			return CV_8UC4;
	} else if (encoding == "bgra8") {
			return CV_8UC4;
	} else if (encoding == "32FC1") {
			return CV_32FC1;
	} else if (encoding == "rgb8") {
			return CV_8UC3;
	} else {
			throw std::runtime_error("Unsupported encoding type");
	}
}